Translation of the 'useState' page (#490)
reactjs/fr.react.dev: src/content/reference/react/useState.md

          checked={liked}
          onChange={handleChange}
        />

>         J'ai aimé

Author: Quel est la plus-value de ce changement ? Est-ce uniquement visuel ?

>         J'ai aimé

Reviewer: Oui, on vise des apostrophes typographiques partout dans les textes affichés. Pour la prose leur post-traitement Markdown le fait pour nous mais pour les légendes de diagrammes et le code ce n'est pas le cas.

1. `a => a + 1` recevra un état en attente valant `43` et va retourner un prochain état valant `44`.
1. `a => a + 1` recevra un état en attente valant `44` et va retourner un prochain état valant `45`.

Il n'y a pas d'autres mises à jour en file d'attente, React stockera donc à la fin `45` comme étant l'état courant.

Par convention, il est commun de nommer l'argument de l'état en attente selon la première lettre du nom de la variable d'état, comme `a` pour `age`. Cependant, vous pouvez également le nommer `prevAge`, ou quelque chose d'autre que vous trouvez plus clair.

En développement, React pourra [appeler deux fois vos mises à jour](#my-initializer-or-updater-function-runs-twice) pour vérifier si elles sont [pures](/learn/keeping-components-pure).

<DeepDive>

#### Est-il toujours préférable d'utiliser une mise à jour ? {/*is-using-an-updater-always-preferred*/}

Vous pourrez peut-être entendre des recommandations vous disant de toujours écrire votre code de cette manière, si l'état que vous mettez à jour est calculé depuis l'état précédent : `setAge(a => a + 1)`. Il n'y a aucun mal à le faire, mais ce n'est pas toujours nécessaire.

> Dans la plupart des cas, il n'y a aucune différence entre ces deux approches. React vérifiera toujours, pour les actions intentionnelles des utilisateurs, que l'état `age` soit mis à jour avant le prochain click. Cela signifie qu'il n'y a aucun risque à ce qu'un gestionnaire de click voit un `age` "obsolète" au début de la gestion des évènements.

Author: Je note !

---

### Mettre à jour des objets et des tableaux dans un état {/*updating-objects-and-arrays-in-state*/}

Vous pouvez insérer des objets et des tableaux dans un état. En React, un état est considéré comme en lecture seule, **vous devez donc *replacer* vos objets existants plutôt que les *muter***. Par exemple, si vous avez un objet `form` dans un état, ne le mutez pas :

```js
// 🚩 Ne mutez pas un objet dans un état comme ceci :
form.firstName = 'Clara';
```

À la place, replacez l'objet entier en en créant un nouveau :

```js
// ✅ Remplacez l'état avec un nouvel objet 
setForm({
  ...form,
  firstName: 'Clara'
});
```

Lire [mettre à jour des objets dans un état](/learn/updating-objects-in-state) et [mettre à jour des tableaux dans un état](/learn/updating-arrays-in-state) pour en savoir plus.

<Recipes titleText="Exemples d'objets et de tableaux dans un état" titleId="examples-objects">

#### Formulaire (objet) {/*form-object*/}

> Dans cet exemple, la variable d'état `form` contient un objet. Chaque champ de saisie possède un gestionnaire de changement qui appelle `setForm` avec le prochain état du formulaire tout entier. La syntaxe de décomposition `{...form}` permet de s'assurer que l'état de l'objet est remplacé, plutôt que muté.

Author: Je suis bien d'accord. Je m'étais en effet basée sur le MDN FR pour cette traduction.

<Solution />

> #### Écrire la logique d'une mise à jour de manière concise avec Immer {/*writing-concise-update-logic-with-immer*/}

Author: Jamais utilisé, mais ils m'ont donné envie :D

  return (
    <>

>       <h1>Liste d'Art</h1>

Author: Compliqué à traduire celui-ci, c'est mieux comme ça !

---

> ### Stocker les informations des rendus précédents {/*storing-information-from-previous-renders*/}

Author: Pourquoi ce FIXME ?

Dans la plupart des cas, vous n'avez pas besoin de ça :

* **Si la valeur dont vous avez besoin peut être totalement calculée à partir des propriétés actuelles ou d'un autre état, [supprimez complètement ces états redondants](/learn/choosing-the-state-structure#avoid-redundant-state)**. Si effectuer de nouveaux calculs trop fréquemment vous dérange, le [Hook `useMemo`](/reference/react/useMemo) peut vous aider.   
* Si vous voulez réinitialiser l'entièreté de l'arbre d'un composant, [passez une `key` différente à votre composant](#resetting-state-with-a-key).
* Si vous le pouvez, mettez à jour tous les états correspondants dans des gestionnaires d'évènements.

Dans de rares autres cas, il existe un modèle que vous pouvez utiliser pour mettre à jour un état selon les états qui ont été rendus jusqu'ici, en appelant une fonction de mise à jour pendant que votre composant est en cours de rendu.

Voici un exemple. Ce composant `CountLabel` affiche une propriété `count` qui lui est passé :

```js CountLabel.js
export default function CountLabel({ count }) {
  return <h1>{count}</h1>
}
```

> Mettons que vous vouliez montrer si le compteur a *augmenté ou diminué* depuis le dernier changement. La propriété `count` ne vous permet pas de savoir ça -- vous avez besoin de garder une trace de sa dernière valeur. Ajoutez la variable d'état `prevCount` afin de garder une trace de sa dernière valeur. Ajoutez une autre variable d'état appelée `trend` qui permet de savoir si le compteur a augmenté ou diminué. Comparez `prevCount` avec `count`, et, s'ils ne sont pas égaux, mettez à jour `prevCount` et `trend`. Vous pouvez maintenant montrer la propriété du compteur courante et *comment elle a changé depuis le dernier rendu*.

Author: Bien vu ! Je n'y avais pas pensé en traduisant ! :open_mouth:

</Sandpack>

Notez que si vous appelez une fonction de mise à jour pendant le rendu, elle doit être dans une condition, telle que `prevCount !== count`, et il doit y avoir un appel dans la condition, tel que `setPrevCount(count)`. Sinon, votre composant effectuera des rendus en boucle jusqu'à crash. De plus, vous pouvez seulement mettre à jour l'état du composant *en cours de rendu* de cette manière. Appeler la fonction de mise à jour d'un *autre* composant pendant le rendu est une erreur. Finalement, votre appel de mise à jour devrait toujours [mettre à jour l'état sans mutation](#updating-objects-and-arrays-in-state) -- cela ne veut pas dire que vous pouvez enfreindre les autres règles des [fonctions pures](/learn/keeping-components-pure).

> Cette méthode peut être difficile à comprendre et est la plupart du temps à éviter. Cependant, c'est toujours mieux que de mettre à jour un état dans un effet. Lorsque vous appelez une fonction de mise à jour pendant le rendu, React va effectuer un autre rendu de ce composant immédiatement après que votre composant se soit terminé grâce à un `return`, et avant d'effectuer un rendu de ses enfants. De cette manière, les enfants n'ont pas besoin d'effectuer deux rendus. Le reste de votre composant va toujours s'effectuer (et le résultat va être jeté). Si votre composant est en-dessous de tous les appels à des Hook, vous devez ajouter un `return;` en avance pour redémarrer plus tôt le rendu.

Author: On peut aussi laisser une issue demandant à React de le préciser dans la doc originale. ;)

---

## Dépannage {/*troubleshooting*/}

> ### J'ai mis à jour l'état, mais le logging m'affiche toujours l'ancienne valeur {/*ive-updated-the-state-but-logging-gives-me-the-old-value*/}

Reviewer: ```suggestion
### J'ai mis à jour l'état, mais je vois toujours l'ancienne valeur {/*ive-updated-the-state-but-logging-gives-me-the-old-value*/}
```

Appeler la fonction de mise à jour **ne change pas l'état dans le code en cours de lancement** :

```js {4,5,8}
function handleClick() {
  console.log(count);  // 0

>   setCount(count + 1); // Demander un nouveau rendu avec 1

Reviewer: ```suggestion
  setCount(count + 1); // Demande un nouveau rendu avec 1
```

  console.log(count);  // Toujours 0 !

  setTimeout(() => {
    console.log(count); // Encore 0 !
  }, 5000);
}
```

> C'est parce que l'[état se comporte comme une snapshot](/learn/state-as-a-snapshot). Mettre à jour l'état déclenche un autre rendu avec la nouvelle variable d'état, mais cela n'affecte pas la variable JavaScript `count` dans le gestionnaire d'évènements en train de s'exécuter.

Reviewer: ```suggestion
C'est parce que l'[état se comporte comme un instantané](/learn/state-as-a-snapshot). Mettre à jour l'état planifie un autre rendu avec la nouvelle valeur d'état, mais n'affecte pas la variable JavaScript `count` dans le gestionnaire d'évènements en train de s'exécuter.
```

---

> ### J'ai mis à jour l'état, mais l'écran ne se met pas à jour {/*ive-updated-the-state-but-the-screen-doesnt-update*/}

Reviewer: ```suggestion
### J'ai mis à jour l'état, mais l'affichage ne se met pas à jour {/*ive-updated-the-state-but-the-screen-doesnt-update*/}
```

> React va **ignorer votre mise à jour si le prochain état est égal à l'état précédent**, en comparant au moyen de [`Object.is`](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/Object/is). Ça arrive la plupart du temps lorsque vous changez un objet ou un tableau directement dans l'état :

Reviewer: ```suggestion
React **ignorera votre mise à jour si le prochain état est égal à l'état précédent**, en comparant au moyen de [`Object.is`](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Global_Objects/Object/is). Ça arrive généralement lorsque vous modifiez directement un objet ou un tableau dans l'état :
```

```js

> obj.x = 10;  // 🚩 Faux : mutation d'un objet existant

Reviewer: ```suggestion
obj.x = 10;  // 🚩 Erroné : mutation d’un objet existant
```

setObj(obj); // 🚩 Ne fait rien
```

> Vous avez muté un objet `obj` existant et vous l'avez re-passé à `setObj`, donc React ignore la mise à jour. Pour résoudre cela, vous devez vous assurer que vous [remplacez toujours les objets et les tableaux dans l'état plutôt que les _muter_](#updating-objects-and-arrays-in-state) :

Reviewer: ```suggestion
Vous avez modifié un objet `obj` existant et vous l'avez passé à `setObj`, donc React ignore la mise à jour (c'est la même référence, le même objet en mémoire). Pour corriger ça, vous devez vous assurer de [toujours _remplacer_ les objets et les tableaux de l'état plutôt que les _modifier_](#updating-objects-and-arrays-in-state) :
```

```js

> // ✅ Correct : créer un nouvel objet

Reviewer: ```suggestion
// ✅ Correct : création d’un nouvel objet
```

---

### J'ai une erreur : "Too many re-renders" {/*im-getting-an-error-too-many-re-renders*/}

> Vous pourriez peut-être avoir une erreur disant : `Too many re-renders. React limits the number of renders to prevent an infinite loop`. La plupart du temps, cela signifie que vous mettez à jour un état de manière non conditionnelle *pendant le rendu*, donc votre composant rentre dans une boucle : rendu, mise à jour de l'état (qui cause un rendu), rendu, mise à jour de l'état (qui cause un rendu), etc. Peu souvent, c'est causé par une erreur en spécifiant le gestionnaire d'évènement : 

Reviewer: ```suggestion
Vous verrez peut-être une erreur disant : `Too many re-renders. React limits the number of renders to prevent an infinite loop` _(« Trop re rendus successifs.  React limite le nombre de rendus pour éviter une boucle infinie », NdT)_. Ça signifie le plus souvent que vous mettez à jour un état de manière inconditionnelle *pendant le rendu*, de sorte que votre composant entre dans une boucle : rendu, mise à jour de l'état (qui déclenche un rendu), rendu, mise à jour de l'état (qui entraîne un rendu), etc. Le plus souvent, ça vient d'une erreur classique de fourniture d'un gestionnaire d'évènement : 
```

Author: "une erreur classique de fourniture d'un gestionnaire d'évènement" => n'est-ce pas un peu verbeux ?

```js {1-2}

> // 🚩 Faux : appelle le gestionnaire pendant le rendu

Reviewer: ```suggestion
// 🚩 Erroné : appelle le gestionnaire pendant le rendu
```

return <button onClick={handleClick()}>Click me</button>

> // ✅ Correct : passe le gestionnaire d'évènement 

Reviewer: ```suggestion
// ✅ Correct : passe le gestionnaire d’évènement 
```

return <button onClick={handleClick}>Click me</button>

> // ✅ Correct : passe une fonction en ligne

Reviewer: ```suggestion
// ✅ Correct : passe une fonction créée à la volée
```

return <button onClick={(e) => handleClick(e)}>Click me</button>
```

> Si vous n'arrivez pas à trouver le cause de cette erreur, cliquez, dans la console, sur la flèche à côté de l'erreur, et regardez dans votre pile JavaScript afin de trouver la fonction de mise à jour responsable de cette erreur.

Reviewer: ```suggestion
Si vous n'arrivez pas à trouver le cause de cette erreur, cliquez dans la console sur la flèche à côté de l'erreur, et examinez votre pile d'appels JavaScript afin d'y repérer la fonction de mise à jour responsable de l'erreur.
```

---

> ### Ma fonction d'initialisation ou ma fonction de mise à jour se lance deux fois {/*my-initializer-or-updater-function-runs-twice*/}

Reviewer: ```suggestion
### Ma fonction d'initialisation ou ma fonction de mise à jour est exécutée deux fois {/*my-initializer-or-updater-function-runs-twice*/}
```

> En [mode Strict](/reference/react/StrictMode), React va appeler certaines de vos fonctions deux fois plutôt qu'une :

Reviewer: ```suggestion
En [mode strict](/reference/react/StrictMode), React appellera certaines de vos fonctions plutôt deux fois qu'une :
```

Autant s'amuser avec l'expression consacrée 😉

```js {2,5-6,11-12}
function TodoList() {

>   // La fonction de ce composant va se lancer deux fois pendant le rendu.

Reviewer: ```suggestion
  // Cette fonction composant sera appelée deux fois par rendu.
```

  const [todos, setTodos] = useState(() => {

>     // Cette fonction d'initialisation va se lancer deux fois pendant le rendu.

Reviewer: ```suggestion
    // Cette fonction d'initialisation sera appelée deux fois par rendu.
```

    return createTodos();
  });

  function handleClick() {
    setTodos(prevTodos => {

>       // Cette fonction de mise à jour va se lancer deux fois à chaque click.

Reviewer: ```suggestion
      // Cette fonction de mise à jour sera appelée deux fois par clic.
```

      return [...prevTodos, createTodo()];
    });
  }
  // ...
```

C'est voulu, et ça ne devrait pas casser votre code.

> Ce comportement, **uniquement présent en développement**, vous aide à [garder vos composants purs](/learn/keeping-components-pure). React utilise le résultat d'un des appels, et ignore le résultat de l'autre. Tant que vos composants, vos fonctions d'initialisation, et vos fonctions de mise à jour sont pures, ça ne devrait pas affecter votre logique. Cependant, si elles sont accidentellement impures, ça vous aide à remarquer vos erreurs. 

Reviewer: ```suggestion
Ce comportement, **uniquement présent en développement**, vous aide à [garder vos composants purs](/learn/keeping-components-pure). React utilise le résultat d'un des appels et ignore le résultat de l'autre. Tant que vos composants, vos fonctions d'initialisation, et vos fonctions de mise à jour sont pures, ça ne devrait pas affecter le comportement. Cependant, si elles sont accidentellement impures, ça vous aide à détecter le problème. 
```

> Par exemple, cette fonction de mise à jour impure mute un tableau dans un état :

Reviewer: ```suggestion
Par exemple, cette fonction de mise à jour impure modifie directement un tableau dans un état :
```

```js {2,3}
setTodos(prevTodos => {

>   // 🚩 Erreur : muter un état

Reviewer: ```suggestion
  // 🚩 Erreur : modification en place de l'état
```

  prevTodos.push(createTodo());
});
```

> Comme React a appelé votre fonction de mise à jour à deux reprises, vous verrez que le `todo` a été appelé deux fois, et vous saurez qu'il y a une erreur. Dans cet exemple, vous pouvez résoudre l'erreur en [remplaçant le tableau, plutôt que de le muter](#updating-objects-and-arrays-in-state) :

Reviewer: ```suggestion
Comme React a appelé votre fonction de mise à jour à deux reprises, vous verrez que la tâche a été ajoutée deux fois, et vous saurez qu'il y a une erreur. Dans cet exemple, vous pouvez corriger l'erreur en [remplaçant le tableau, plutôt que de le modifier](#updating-objects-and-arrays-in-state) :
```

```js {2,3}
setTodos(prevTodos => {

>   // ✅ Correct : remplacer avec le nouvel état

Reviewer: ```suggestion
  // ✅ Correct : remplacement par un nouvel état
```

  return [...prevTodos, createTodo()];
});
```

> Maintenant que cette fonction de mise à jour est pure, l'appeler une autre fois ne fait aucune différence de comportement. C'est pour cette raison que l'appel à deux reprises de React vous aide à trouver des erreurs. **Seulement les composants, les fonctions d'initialisation et les fonctions de mise à jour doivent être pures**. Les gestionnaires d'évènements n'ont pas besoin d'être purs, donc React ne les appellera jamais deux fois.

Reviewer: ```suggestion
Maintenant que cette fonction de mise à jour est pure, l'appeler une fois de plus n'entraîne aucune différence de comportement. C'est en cela que le double appel par React vous aide à détecter les problèmes. **Seuls les composants, les fonctions d'initialisation et les fonctions de mise à jour doivent être purs**. Les gestionnaires d'évènements n'ont pas besoin d'être purs, aussi React ne les appellera jamais deux fois.
```

> Lire [garder les composants purs](/learn/keeping-components-pure) pour en savoir plus.

Reviewer: ```suggestion
Consultez [Garder les composants purs](/learn/keeping-components-pure) pour en savoir plus.
```

---

> ### J'essaye de mettre à jour un état avec une fonction, mais elle est appelée à la place {/*im-trying-to-set-state-to-a-function-but-it-gets-called-instead*/}

Reviewer: ```suggestion
### J'essaie de placer une fonction dans un état, mais elle est appelée directement {/*im-trying-to-set-state-to-a-function-but-it-gets-called-instead*/}
```

}
```

> Comme vous passez une fonction, React part de principe que `someFunction` est une [fonction d'initialisation](#avoiding-recreating-the-initial-state), et que `someOtherFunction` est une [fonction de mise à jour](#updating-state-based-on-the-previous-state), et donc essaye de les appeler et de stocker leurs résultats. Pour réellement *stocker* une fonction, dans les deux cas, vous devez placer en amont `() =>`. Dans ce cas, React stockera les fonctions que vous passez.

Reviewer: ```suggestion
Comme vous passez une fonction, React part du principe que `someFunction` est une [fonction d'initialisation](#avoiding-recreating-the-initial-state), et que `someOtherFunction` est une [fonction de mise à jour](#updating-state-based-on-the-previous-state) ; il va donc les appeler pour stocker leurs résultats. Afin de *stocker* effectivement une fonction, dans les deux cas vous devrez la préfixer par `() =>`. Dans ce cas, React stockera les fonctions que vous passez.
```
